scrutiny/server/device/links/rtt_link.py

```diff
@@ -0,0 +1,177 @@
+
+    def __init__(self, config: LinkConfig):
+        self.port = None
+        self.validate_config(config)
+        self._initialized = False
+        self.logger = logging.getLogger(self.__class__.__name__)
+        
+
+        self.config = cast(RttConfig, {
+            'target_device': str(config['target_device']),
+            'write_timeout_delay' : float(config.get('write_timeout_delay', 0)),
+            'jlink_interface' : str(config.get('jlink_interface','swd'))
+        })
+
+    def get_config(self) -> LinkConfig:
+        return cast(LinkConfig, self.config)
+
+    def initialize(self) -> None:
+        """ Called by the device Handler when initiating communication. Should reset the channel to a working state"""
+        target_device = str(self.config['target_device'])
+        jlink_interface = self.get_jlink_interface(self.config['jlink_interface'])
+
+        self.port = pylink.JLink()
+
+        self.port.open()
+        if self.port.opened():
+            self.logger.debug("J-Link opened: " + str(self.port.product_name))
+        else:
+            self.logger.debug("J-Link not opened." )
+
+        self.port.set_tif(jlink_interface)
+        self.port.connect(target_device)
+        self.port.rtt_start(None)
+
+        if self.port.target_connected():
+            self._initialized = True
+            self.logger.debug("J-Link connected: " + str(target_device))
+        else:
+            self._initialized = False
+            self.logger.debug("J-Link not connected: " + str(target_device))      
+
+    def destroy(self) -> None:
+        """ Put the comm channel to a resource-free non-working state"""
+        if self.port is not None:
+            if self.port.opened():
+                self.port.close()
+        self._initialized = False
+
+    def operational(self) -> bool:
+        """ Tells if this comm channel is in proper state to be functional"""
+        if self.port is None:
+            return False
+        return self.port.connected() and self.initialized() and self.port.opened() and self.port.target_connected()
+
+    def read(self, timeout:Optional[float] = None) -> Optional[bytes]:
+        """ Reads bytes in a blocking fashion from the comm channel. None if no data available after timeout"""
+        if not self.operational():
+            return None
+        
+        assert self.port is not None    # For mypy
+        data:bytes = None
+        try:
+            bytesArray = self.port.rtt_read(0, 1024)
+            data = bytes(bytesArray)
+        except Exception:
+            self.logger.debug("Cannot read data.")
+            self.port.close()
+        return data
+
+
+    def write(self, data: bytes) -> None:
+        """ Write data to the comm channel."""
+        if self.operational():
+            assert self.port is not None    # For mypy
+            try:
+                total_number_bytes = len(data)
+                number_byte_written = 0
+                timestamp = time.monotonic()
```

Comment:
use time.perf_counter for precise timer.   monotonic timer has 16ms resolution on windows.